### ISSUE TITLE
Add a way to filter posts that contain images with no alt text

### ISSUE THREAD

**Issue by Yekrats** [mastodon/mastodon#29496]
### Pitch

It would be nice if there would be some way of filtering out posts from one's feed that do not contain alt-text for images. Ostensibly, this would be a fairly simple query to write for the back-end. I would imagine it would be a simple check-box for the front-end in the Filter tab of the Preferences page.

### Motivation

This would help mitigate the spread of images across the Fediverse without alt text.

**Comment by Cassolotl** [comment]
Just curious - @Yekrats, do you use a screen reader?

**Comment by Yekrats** [comment]
@Cassolotl No, I do not.

**Comment by Cassolotl** [comment]
Okay. :) My understanding is that alt text is not always useful, such as when it's obvious what's in the photo from the main body of the post, e.g. "Look at this cute photo of my cat wearing a little cowboy hat!" And in those situations, adding alt text is actually anti-helpful because it just repeats something and wastes time.

Having said that, I don't use a screen reader either, so we are just two sighted people speculating about what visually impaired people might want.

**Comment by YourAutisticLife** [comment]
> Okay. :) My understanding is that alt text is not always useful, such as when it's obvious what's in the photo from the main body of the post, e.g. "Look at this cute photo of my cat wearing a little cowboy hat!" And in those situations, adding alt text is actually anti-helpful because it just repeats something and wastes time.
> 
> Having said that, I don't use a screen reader either, so we are just two sighted people speculating about what visually impaired people might want.

This is precisely the behavior that this filter aims to disfavor. We cannot prevent people from doing the above. However, what the person who posts the hypothetical picture above should do is make the text of the post say:

"Look at this cute photo."

And set the alt text of the image to "My cat wearing a little cowboy hat."

Yes, it does require some thinking on the part of the person who posts. That's what this change aims to promote: that people think about what they are posting so that people who use screen readers are not subject to pictures that don't have alt text.

And yes, this is a change that I've had to do for myself. I did not use to care about alt text, but now I do. So I've changed my posting habits to put alt text to my images. In some cases, this means cutting and pasting the textual content of the post as alt text. I don't feel great about this, but this, at least, don't leave people with screen readers wondering. 

This is a fairly low bar to meet.

Also, I'll say that the textual content of a post may, sometimes, become divorced from the image that it accompanies. If I go to the "Media" tab of my profile, I see there all the images that I posted to Mastodon. If I click on an image, I get it blown up, with its alt text, but without the text of the post in which it was published.

**Comment by benaryorg** [comment]
> do you use a screen reader?

I would like to mention that alt-text is an accessibility tool *not just* for people with impaired eyesight, although that is one of the primary concerns for people implementing it AFAIK.
Some pictures/videos/etc. just have a lot going on and are generally overwhelming, and alt-text can tell people what the actual content is.
Alt-text is often used to name people or things in there which others may not know like regional food which cannot be properly recognized without alt-text, or the picture of a person that mainstream media may be able to identify as Celebrity Foo, but that someone outside that bubble may not know.
Videos with flash images or such that cannot be watched by some people lest they *really* wanna have a seizure right now are also inaccessible without alt-text.

These are all different groups that run into different issues when people do not provide alt-text.
In a way it's media that just goes "hey look at this thing" without any *thing* for them to look at.

Yes, the primary goal of alt-text is to be accommodating to people who cannot otherwise perceive the media in question, and of course we should do that just because accommodating other people is the right thing to do.
However, for those that need extra convincing, how often have you been in a public place and there was an audio file or video in your timeline that you couldn't properly watch because you didn't have headphones with you and don't want to annoy the people around you?
How often have you been in a building with really bad reception and media has taken forever to load, but the alt-text was there right away so you could decide whether or not to wait for the media to load based on whether you thought it was interesting based on the alt-text?
Accessibility helps everyone.
However, I would like to make clear that helping people without issues perceiving the media should come second to helping the people who cannot participate otherwise at all.
I'm just here to name a few examples that make non-alt-texted media problematic in general to justify such a toggle.

**Comment by KreerC** [comment]
Hello, screen reader user and accessibility consultant here.

First of all, thanks for bringing this to public attention. @YourAutisticLife has already explained very well why leaving the alt text empty is never a good idea. Blind people will generally not know oder likely understand that the image contents are already described in the text.

Regarding the actual suggestion, I feel a bit conflicted about whether this is a good feature. **I personally would very likely not opt to use it**, for the following reasons:

- Having the filter could encourage people to actively discriminate, telling people that they can simply hide images without alt text through a simple setting. Seeing that Mastodon Web is by far the most used fediverse client, implementing this might trigger several political discussions where disabled users are very likely to be in the minority and in a rather weak position.
- Missing alt text can be a reminder to always post *with* alt text itself
- Sometimes, people genuinely forget to add an alt text. They will probably start to wonder why their posts get much less interaction. They might get deterred from posting on the fediverse altogether.
- As a blind person, I will lose the full picture of what is actually happening in the world and what is supposed to be on my timeline. Especially as more and more politicians are using the platform, using this filter might have a severe impact on what sources I will notice that even exist on a topic.
- More and more screen readers introduce AI-based image description features in cases where no alt text is provided with the content. This can never replace a human caption by the poster, but in many cases it will be good enough for the user to make some more sense of the posts.

However, that's not the full picture. Arguments *for* such a filter might be:

- The filter could cause more users to provide alt text with their media, especially if would a large-scale campaign surrounding it that would tell people to turn it on. 
- It could *potentially* save a huge amount of time and frustration for blind users casually browsing a timeline. 

In the end, this will require quite a bit of political debate. I think that the negatives far outweigh the pros of this feature, at least from my POV: that of a blind, but politically very interested person. Before implementing it, perhaps it would be a good idea to test it through another very popular client and see how the network reacts. More people with disabilities should be brought into the discussion as well.

**Comment by Cassolotl** [comment]
I appreciate the comments from an accessibility consultant, but this is still a discussion entirely between people who are not affected by the feature in an accessibility way...

Does anyone know any people who use screen readers or who primarily view images as alt text for access reasons, who we can invite to comment here?

**Comment by KreerC** [comment]
As I already said in the post, I am a blind screen reader user **and** an accessibility consultant.

**Comment by Cassolotl** [comment]
> As I already said in the post, I am a blind screen reader user **and** an accessibility consultant.

Somehow I missed that, I'm so sorry! I'll edit my snarky post. :D

**Comment by deborahgu** [comment]
Hi,

Self-identification

* accessibility consultant
* sighted
* disabled assistive tech user
* but not screen reader or braille display user
* sometimes switch to alt text with images off anyway, for both non-vision disabilities and age-related vision issues

This would be a great feature, honestly; I would sometimes use it and definitely tell other people about it.

In answer to the comment above from @Cassolotl:

> My understanding is that alt text is not always useful

Unfortunately, if there's no alt at all (rather than blank alt), screen readers and braille displays default to reading something noisy and rarely-informative, usually the filename. I _almost_ would rather there'd be a feature to mark images as decorative (`alt=""`) so if the image is fully described in text people can just click a box to eliminate noise to the alt's consumer, but honestly it's hard enough getting full time content creators to understand what should be marked with `alt=""`, and on a microblogging site it's just asking for trouble.

(Note: this does not mean absent alt should be replaced with blank alt; blank alt means a conscious decision has been made that this image is pure filler or is described in text elsewhere.)

**Comment by deborahgu** [comment]
Also, I understand why you're asking for people to self-identify and I'm not sure I wouldn't do the same, but, counterpoint, asking disabled people to out themselves online in order to have their opinions count is inherently problematic. A github account is (for many of us) tied to a personal or professional identity, and it's not always reasonable or safe to disclose disability status or assistive tech use in that context.

**Comment by rkingett** [comment]
As a screen reader user, I want this more than anything! I'd like to just get rid of all images on my timeline/audio/video/media that don't have a description! Far better than having people whine because we suggest they add it.

**Comment by hopesthoughts** [comment]
Here's a bit of an idea, just as an extra option, add a way to filter out all images. Don't make it required or mandatory at all. I just don't prefer to come across them, but that doesn't mean someone else won't.

**Comment by hopesthoughts** [comment]
As a bit of an addendum, I do use a screen reader, but not an accessibility expert in any way! On the part about alt text wasting time, it definitely can. I think if your just wanting to post something, a few words is fine.
     It's a bit different if your going to charge me something from your own site. Of course then I want to know all about it.
     I still think filtering out all images, as a preference would be cool. Of course no one would want to implement that.

**Comment by MadokVaur** [comment]
> ### Pitch
<snip>> This would help mitigate the spread of images across the Fediverse without alt text.

Now that this has dragged on for a week with the usual digression to the screen-reader / accessibility issue, let's return to the original post:

> This would help mitigate the spread of images across the Fediverse without alt text.

How exactly is that going to happen unless Mastodon as a distribution takes it upon itself to become a global Hall Monitor and post a nasty-gram back to anyone who does *not* use alt-text and who *does* have their posts filtered out by each individual person invoking this filter?

OK: the person *using* the filter sees no images without alt-text. Their loss, one might say. Are they going to trigger the nasty-gram themselves? Identifiable as being from them?

Or does Mastodon as a distribution at large implement a pretty draconian behavioral policy change, globally?

So how does the Original Poster ever learn anything?

People using the filter don't see images that offend them

The people (or bots, it must be said) posting the offending images have learned nothing

**Comment by rkingett** [comment]
I'd also be happy with a filter to remove all media, no matter what kind of media attachment it is, video, audio, et and just have text and links for a week or a day or something.

**Comment by deborahgu** [comment]
> Now that this has dragged on for a week with the usual digression to the screen-reader / accessibility issue, let's return to the original post:

It's weird to think of accessibility as a digression on this of all questions. Regardless of the social issue, images without alt text provide noise that many screen reader and braille display users don't want, and giving those users an easy way to filter those posts would be an accessibility win for them. Images without alt text are _noisy_: they read a filename. On a long form page, knowing you missed something is important, so that feature is valuable. On a social media post where an undescribed image is often the lion's share of the post, some users will consider the noise unacceptable.

Whatever social engineering motivation @Yekrats had in making the feature request is irrelevant to the fact that this would improve the accessibility of the site for some disabled users.

**Comment by blaiddwyn** [comment]
i've already requested this feature from Tapbot (developers of Ivory, which has an alt text reminder setting)… i always read alt text when available, cannot read text from images, and see many images of text without alt text or with inadequate alt text or description

**Comment by MadokVaur** [comment]
> It's weird to think of accessibility as a digression on this of all questions. Regardless of the social issue, images without alt text provide noise that many screen reader and braille display users don't want, and giving those users an easy way to filter those posts would be an accessibility win for them. 

You ignore my question, and my point

Exactly and with specificity: how do you go about making this a teachable moment and educate those who (for whatever reason) do not add alt-text to images they post?

The fact that there are those who use screen readers and Braille display readers is understood

How do you propose to go about reaching and educating those who do *not* use alt-text?

What is the mechanism of instruction?

**Comment by deborahgu** [comment]
No, I didn't ignore your point, I clarified that it's not relevant. I don't care if it's a teachable moment; that was @Yekrats's motivation with the original issue. My point is that _it's a good feature_. It's not a good feature to educate, it's a good accessibility feature, full stop.

**Comment by MadokVaur** [comment]
Helping people learn why they should be *using* alt-text is not relevant?

Really?

It (hiding posts lacking alt-text) affects only what the person *using* the filters sees

It does nothing to help the Original Poster learn why they should be using alt-text and it's not going to result in any sort of teachable moment for those who don't

So you're not interested in education and change, so much as an endless stream of opportunities for performative allyship by those complaining, directed at those you aren't interested in helping to learn

Got it

10-4 and out

**Comment by deborahgu** [comment]
No, I'm a disabled person who is interested in my tools being more accessible, and doesn't think the purpose of accessible software is to educate. I am confused by what it is you think I am saying but you are clearly much too angry at an invisible person to have a two-sided conversation about accessibility features, so I wish you peace with whatever has angered you.

**Comment by ryan-p-randall** [comment]
For whatever it's worth, two separate feature proposals remains a possibility.

This proposed one is aimed at individual users who want to control their own feed. Another feature could be proposed that would aspire to educating all users.

**Comment by rkingett** [comment]
So I don't think filtering out just images would work because audio/video gets interpreted the same as an image by Mastodon I believe, so this would filter out audio and or video as well if we have an option to hide all media/images.

**Comment by ChasBelov** [comment]
I do sometimes see alt text on videos.

**Comment by Scobiform** [comment]
Just make alt-text mandatory. Solves all problems.

**Comment by rkingett** [comment]
Given the culture of new people coming to the Fedi being what it is, if we make alt text mandatory, then people will just abuse the alt text feature to the point where it becomes double the work to assess the quality of alt text.

**Comment by Scobiform** [comment]
> Given the culture of new people coming to the Fedi being what it is, if we make alt text mandatory, then people will just abuse the alt text feature to the point where it becomes double the work to assess the quality of alt text.

I get your point. But it should hurt people a little bit more not to set alt-text. Alt-Text is not optional, its is mandatory for severely disabled people like me. Give users some tools to generate alt-text and a Tutorial why it is important...

**Comment by FurblandChannel** [comment]
I second this. I think it would also be nice to simply set it up so posts without alt text are not allowed to trend. This would be a nice balance between requiring it and not requiring it.

**Comment by ChasBelov** [comment]
While you're waiting for this feature, the following CSS seems to hide such posts:

#mastodon div.status:has(img:not([alt])), 
#mastodon div.detailed-status:has(img:not([alt])) {
    display:none !important;
}

Code subject to breakage due to Mastodon code changes.

**Comment by Diwannist** [comment]
Adding a user preference that mute toots that contain images without alt text may be good.

**Comment by Cassolotl** [comment]
> In recent years, there has been a rush on the internet to supply image descriptions and to call out those who don’t. This may be an example of community accountability at work, but it’s striking to observe that those doing the most fierce calling out or correcting are sighted people. Such efforts are largely self-defeating. I cannot count the times I’ve stopped reading a video transcript because it started with a dense word picture. Even if a description is short and well done, I often wish there were no description at all. Get to the point, already! How ironic that striving after access can actually create a barrier. When I pointed this out during one of my seminars, a participant made us all laugh by doing a parody: “Mary is wearing a green, blue, and red striped shirt; every fourth stripe also has a purple dot the size of a pea in it, and there are forty-seven stripes—”

An excerpt from "[Against Access](https://audio.mcsweeneys.net/transcripts/against_access.html)", an essay by John Lee Clark, just under 5,000 words.

**Comment by rkingett** [comment]
Any person should know that doesn't represent all us blind people. But of course, to make the point that we shouldn't have anything we want, that article is used without understanding the article in full. Fantastic. But of course, he reminds me of the very same blind people that advocate against knowing anything about, say, racial discrimination, because disability justice matters exclusively without intersectionality to these people.

**Comment by VorEver** [comment]
Alternate text isn't *only* for screen reader users.  Some people have "face blindness" and cannot recognize people in photos/images. Other people have other issues with images.

In my case, I was sight-restored as a teenager. The part of my brain that manages recognition didn't develop properly because, prior to that surgery, I could reach farther than I could actually see. Sometimes, to understand what an image is, I need to be told what I'm looking at.  

I do use a screen reader, but not for everything all the time. Sometimes, I copy the image into a reply and add alt-text when I want to share it with my followers. I never boost if there's an image without alt-text, and tbh, I'd love to filter those baddies out.

To be fair, filtering doesn't mean I'm not seeing that the posts are there; I'm seeing a "this is filtered" instead of the content. I have the option to display it if I choose.  Also, my instance filters some instances, and I can still opt to see the filtered content (profiles, notifications, etc.).

I think this is the part folks who are against this are missing.

**Comment by madeindex** [comment]
I made a post about it and then found this issue was already created.
A lot of people seem to support this feature:
https://mastodon.social/@madeindex/116381133582413117
https://lemmy.world/post/45437791
https://lemmy.world/post/45437770
https://www.reddit.com/r/MadeInDex/comments/1shvzyt/the_fediverse_alttext_issue/
https://www.reddit.com/r/Mastodon/comments/1shw77q/the_fediverse_alttext_issue/

Here is the some of the supporting feedback I collected

Users are are having to rely on their own workarounds:
https://mastodon.cisti.org/@GustavinoBevilacqua/116381174380814021

Some clients are already marking those posts as a workaround:
https://mstdn.social/@MostlyTato/116381261924317153

Additionally it seems some people would like the option to decide where to see alt text and where not:
https://dragonscave.space/@menelion/116381208000692644

**Comment by shaedrich** [comment]
> I do use a screen reader, but not for everything all the time. Sometimes, I copy the image into a reply and add alt-text when I want to share it with my followers. I never boost if there's an image without alt-text, and tbh, I'd love to filter those baddies out.

@VorEver Maybe, similarly to the "remind me to add ALT text to my images" feature, we need a "remind when I attempt to repost/quote a toot containing images without ALT text" feature 🤔 

> Okay. :) My understanding is that alt text is not always useful, such as when it's obvious what's in the photo from the main body of the post, e.g. "Look at this cute photo of my cat wearing a little cowboy hat!" And in those situations, adding alt text is actually anti-helpful because it just repeats something and wastes time.

@Cassolotl It might be, but the toot only paints this picture in the minds of people:

<img width="358" height="271" alt="Image" src="https://github.com/user-attachments/assets/9baa3ceb-f72b-4f95-87dd-def901f5312b" />

but, actually, it can be

| this | or this |
| ---- | ------- |
| <img width="474" height="316" alt="Image" src="https://github.com/user-attachments/assets/f1f72bce-8d16-43d5-8b60-a1c52f269a26" /> | <img width="474" height="265" alt="Image" src="https://github.com/user-attachments/assets/1a34af86-ad9b-4d82-9e54-e7a3470d659d" /> |

> > In recent years, there has been a rush on the internet to supply image descriptions and to call out those who don’t. This may be an example of community accountability at work, but it’s striking to observe that those doing the most fierce calling out or correcting are sighted people. Such efforts are largely self-defeating. I cannot count the times I’ve stopped reading a video transcript because it started with a dense word picture. Even if a description is short and well done, I often wish there were no description at all. Get to the point, already! How ironic that striving after access can actually create a barrier. When I pointed this out during one of my seminars, a participant made us all laugh by doing a parody: “Mary is wearing a green, blue, and red striped shirt; every fourth stripe also has a purple dot the size of a pea in it, and there are forty-seven stripes—”
> 
> An excerpt from "[Against Access](https://audio.mcsweeneys.net/transcripts/against_access.html)", an essay by John Lee Clark, just under 5,000 words.

Different people have different tastes and requirements. We cannot make anyone happy with just one ALT text that should fit all. Some people, especially those, who had been sighted but aren't anymore, know how a cat looks like but they might like to know what kind of cat it is.
